opm/simulators/wells/StandardWell_impl.hpp

```diff
@@ -2263,6 +2240,91 @@ namespace Opm
+    void
+    StandardWell<TypeTag>::
+    updateProductivityIndex(const Simulator& ebosSimulator,
+                            const WellProdIndexCalculator& wellPICalc,
+                            WellState& well_state,
+                            DeferredLogger& deferred_logger) const
+    {
+        auto fluidState = [&ebosSimulator, this](const int perf)
+        {
+            const auto cell_idx = this->well_cells_[perf];
+            return ebosSimulator.model()
+               .cachedIntensiveQuantities(cell_idx, /*timeIdx=*/ 0)->fluidState();
+        };
+
+        auto isInjecting = [this](const typename StandardWell<TypeTag>::FluidState& fs,
+                                  const int perf) -> bool
+        {
+            const auto cell_press = this->extendEval(this->getPerfCellPressure(fs));
+            const auto conn_press = this->getBhp() + this->perf_pressure_diffs_[perf];
+            const auto drawdown   = cell_press - conn_press;
+
+            return drawdown.value() < 0.0;
+        };
+
+        const int np = this->number_of_phases_;
+        auto setToZero = [np](double* x) -> void
+        {
+            std::fill_n(x, np, 0.0);
+        };
+
+        auto addVector = [np](const double* src, double* dest) -> void
+        {
+            std::transform(src, src + np, dest, dest, std::plus<>{});
+        };
+
+        const auto allow_xflow = this->getAllowCrossFlow()
+            || this->openCrossFlowAvoidSingularity(ebosSimulator);
+
+        auto* wellPI = &well_state.productivityIndex()[this->index_of_well_*np + 0];
+        auto* connPI = &well_state.connectionProductivityIndex()[this->first_perf_*np + 0];
+
+        setToZero(wellPI);
+
+        const auto& allConn = this->well_ecl_.getConnections();
+        const auto  nPerf   = allConn.size();
+        for (auto allPerfID = 0*nPerf, subsetPerfID = 0*nPerf; allPerfID < nPerf; ++allPerfID) {
+            if (allConn[allPerfID].state() == Connection::State::SHUT) {
+                continue;
+            }
+
+            std::vector<EvalWell> mob(num_components_, {numWellEq_ + numEq, 0.0});
+            getMobility(ebosSimulator, static_cast<int>(subsetPerfID), mob, deferred_logger);
+
+            const auto& fs = fluidState(subsetPerfID);
+
+            auto connPICalc = [&wellPICalc, allPerfID](const EvalWell& mobility) -> double
+            {
+                return wellPICalc.connectionProdIndStandard(allPerfID, mobility.value());
+            };
+
+            setToZero(connPI);
+            if (! isInjecting(fs, subsetPerfID)) {  // Production or zero flow rate
+                if (allow_xflow || this->isProducer()) {
+                    this->computeConnLevelProdInd(fs, connPICalc, mob, connPI);
+                }
+            }
+            else {              // Connection injects into reservoir
+                if (allow_xflow || this->isInjector()) {
+                    this->computeConnLevelInjInd(fs, connPICalc, mob,
```

Comment:
Passing connPICalc as a connIICalc argument, is that ok?

Reply:
> Passing `connPICalc` as a `connIICalc` argument, is that ok?

I suppose it might be confusing, but yes&mdash;in this context this is okay.  The calculator here is mainly responsible for incorporating logarithmic scaling factors if needed.